### PR TITLE
ENH: DBSCAN integration via oneapi iface

### PR DESCRIPTION
**Pull request by samir-nasibli**
# Description
DBSCAN integration via oneapi iface
* DBSCAN onedal4py impl
* DBSCAN sklearnex dispathcer
* tests:
  * onedal4py
  * sklearnex

# TODO
* currently adding dbscan into main branch. After daal4py|cpu vs sklearnex|cpu measurements will decide to move preview module or not.
* based on this branch I will also add SPMD interface in separate PR -> #1417

### CODE REVIEW

File: onedal/cluster/dbscan.cpp [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+#ifdef ONEDAL_DATA_PARALLEL_SPMD
+    ONEDAL_PY_INSTANTIATE(init_compute_ops, sub, policy_list_spmd, task_list);
+#else // ONEDAL_DATA_PARALLEL_SPMD
+    ONEDAL_PY_INSTANTIATE(init_compute_ops, sub, policy_list, task_list);
+#endif // ONEDAL_DATA_PARALLEL_SPMD
+
+    ONEDAL_PY_INSTANTIATE(init_compute_result, sub, task_list);
+}
+
+} // namespace oneapi::dal::python
```
Comment (Alexsandruss) [inline onedal/cluster/dbscan.cpp:154]: No new line at end of file

File: onedal/cluster/dbscan.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+            "epsilon": float(self.eps),
+            "mem_save_mode": False,
+            "result_options": "core_observation_indices|responses",
+        }
+
+    def _fit(self, X, y, sample_weight, module, queue):
+        policy = self._get_policy(queue, X)
+        X = check_array(X, accept_sparse="csr", dtype=[np.float64, np.float32])
+        ww = make2d(sample_weight) if sample_weight is not None else None
+        XX = make2d(X)
```
Comment (Alexsandruss) [inline onedal/cluster/dbscan.py:75]: Not usual name for modified X
```suggestion
        Xm = make2d(X)
```
```suggestion
        X_ = make2d(X)
```

**Review by Alexsandruss — CHANGES_REQUESTED** [review]

File: sklearnex/cluster/tests/test_dbscan.py [hunk 3]
```diff
@@ -24,7 +24,8 @@ def test_sklearnex_import():
 
     X = np.array([[1, 2], [2, 2], [2, 3], [8, 7], [8, 8], [25, 80]])
     dbscan = DBSCAN(eps=3, min_samples=2).fit(X)
-    assert "daal4py" in dbscan.__module__
+    # assert "daal4py" in dbscan.__module__
+    assert "sklearnex" in dbscan.__module__
```
[the remark this answers is not in the document]
Comment (samir-nasibli) [reply sklearnex/cluster/tests/test_dbscan.py:28]: currently this not in preview, so there is no need for checking daal4py module.

File: onedal/cluster/dbscan.py (hunk 2 above)
Comment (samir-nasibli) [reply onedal/cluster/dbscan.py:75]: PR is progress, refactoring will be done.

**Review by Alexsandruss — DISMISSED** [review]
Perf. comparison to daal4py and sklearn main branch tests result is required before merge.

**Comment by samir-nasibli** [comment]
> Perf. comparison to daal4py and sklearn main branch tests result is required before merge.

Thank you! Again. This PR in progress.  After feature impl will be done all perf tests and comparison analysis for leaving it on main branch or on preview. After that will be done refactoring.

**Comment by samir-nasibli** [comment]
Job: http://intel-ci.intel.com/ee4030be-972c-f138-93c5-a4bf010d0e2e.

**Comment by samir-nasibli** [comment]
GPU CI Job: http://intel-ci.intel.com/ee427fc0-98c0-f1e5-ad74-a4bf010d0e2e. Specifically for sklearnex/daal4py tests/examples on GPU.

File: onedal/cluster/tests/test_dbscan.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+            _test_dbscan_big_data_numpy_gen(
+                queue,
+                eps=eps,
+                min_samples=min_samples,
+                metric=metric,
+                use_weights=use_weights,
+            )
+
+
+@pytest.mark.parametrize("metric", METRIC)
```
Comment (KulikovNikita) [inline onedal/cluster/tests/test_dbscan.py:114]: Don't think we can use any metric different from euclidean

**Comment by ethanglaser** [comment]
> GPU CI Job: http://intel-ci.intel.com/ee427fc0-98c0-f1e5-ad74-a4bf010d0e2e. Specifically for sklearnex/daal4py tests/examples on GPU.

Looks like 3 sklearnex sklearn test failures (on GPU) and test_monkeypatch fail for recipe testing that may require some changes. For now, can just check PVC jobs as DG1 deselected tests are not updated and device issue on recipe for DG1.

**Review by KulikovNikita — APPROVED** [review]

File: onedal/cluster/tests/test_dbscan.py (hunk 4 above)
Comment (samir-nasibli) [reply onedal/cluster/tests/test_dbscan.py]: In any way I think it is better to leave it as is, at least as a useful self doc testcase.

File: sklearnex/cluster/tests/test_dbscan.py [hunk 5]
```diff
@@ -16,15 +16,21 @@
...
 from numpy.testing import assert_allclose
 
 
-def test_sklearnex_import():
+# TODO:
+# adding this parameterized testing
+# somehow breaks other test with preview module patch:
+# sklearnex/tests/test_monkeypatch.py::test_preview_namespace.
+# @pytest.mark.parametrize("dataframe,queue", get_dataframes_and_queues())
+def test_sklearnex_import_dbscan():
```
Comment (samir-nasibli) [inline sklearnex/cluster/tests/test_dbscan.py:28]: Will be addressed on separate PR.

**Comment by samir-nasibli** [comment]
GPU CI Job: http://intel-ci.intel.com/ee45c698-687e-f165-84a9-a4bf010d0e2e. Specifically for sklearnex/daal4py tests/examples on GPU.

File: onedal/cluster/tests/test_dbscan.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    generator = np.random.RandomState()
+    table_size = (samples_number, sample_dimension)
+    return generator.uniform(low=low, high=high, size=table_size), generator.uniform(
+        size=samples_number
+    )
+
+
+def check_labels_equals(left_labels: np.ndarray, right_labels: np.ndarray) -> bool:
+    if left_labels.shape != right_labels.shape:
+        raise Exception("Shapes not equals")
```
Comment (ethanglaser) [inline onedal/cluster/tests/test_dbscan.py:38]:
```suggestion
        raise Exception("Shapes not equal")
```

File: onedal/cluster/tests/test_dbscan.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    if len(left_labels.shape) != 1:
+        raise Exception("Shapes size not equals 1")
+    if len(set(left_labels)) != len(set(right_labels)):
+        raise Exception("Clusters count not equals")
```
Comment (ethanglaser) [inline onedal/cluster/tests/test_dbscan.py:42]:
```suggestion
        raise Exception("Cluster counts not equal")
```

File: sklearnex/cluster/dbscan.py [hunk 8]
```diff
@@ -15,4 +15,167 @@
...
+            "leaf_size": self.leaf_size,
+            "p": self.p,
+            "n_jobs": self.n_jobs,
+        }
+        self._onedal_estimator = onedal_DBSCAN(**onedal_params)
+
+        self._onedal_estimator.fit(X, y=y, sample_weight=sample_weight, queue=queue)
+        self._save_attributes()
+
+    def _onedal_cpu_supported(self, method_name, *data):
```
Comment (Alexsandruss) [inline sklearnex/cluster/dbscan.py:98]: Patching tracking is required for non-preview functionality, for example: [kNN](https://github.com/intel/scikit-learn-intelex/blob/master/sklearnex/neighbors/common.py#L143).

**Review by Alexsandruss — CHANGES_REQUESTED** [review]

File: sklearnex/cluster/dbscan.py (hunk 8 above)
Comment (Alexsandruss) [reply sklearnex/cluster/dbscan.py:98]: Also, `cpu` and `gpu -supported` methods are equal and can be merged even simpler than in kNN

Comment (samir-nasibli) [reply sklearnex/cluster/dbscan.py:98]: Please be kind provide this kind of comments before.
This is non-necessary comment that will patched in the next PR.
